File: app/src/main/java/com/example/pocedex/presentation/PokemonListAdapter.java
package com.example.pocedex.presentation;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;

import com.example.pocedex.data.Pokemon;
import com.example.pocedex.databinding.PokemonItemBinding;

import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

class PokemonListAdapter extends RecyclerView.Adapter<PokemonListAdapter.ViewHolder> {

    private List<Pokemon> pokemons;
    private ItemClickListener clickListener;


    PokemonListAdapter(Context context, List<Pokemon> pokemons) {
        this.pokemons = pokemons;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int ViewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        PokemonItemBinding binding = PokemonItemBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(PokemonListAdapter.ViewHolder holder, int position) {
        Pokemon pokemon = pokemons.get(position);
        holder.binding.setPokemon(pokemon);
    }

    @Override
    public int getItemCount() {
        return pokemons.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        PokemonItemBinding binding;

        ViewHolder(View view) {
            super(view);
            binding = DataBindingUtil.bind(view);

            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (clickListener != null) clickListener.onItemClick(view, getAdapterPosition());
        }
    }

    public Pokemon getPokemon(int id) {
        return pokemons.get(id);
    }

    public void setClickListener(PokemonListAdapter.ItemClickListener itemClickListener) {
        this.clickListener = itemClickListener;
    }

    public interface ItemClickListener {
        void onItemClick(View view, int position);
    }
}

File: app/src/main/java/com/example/pocedex/presentation/PokemonListAdapter.kt
package com.example.pocedex.presentation

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.pocedex.data.Pokemon
import com.example.pocedex.databinding.FavoriteGallaryBinding
import com.example.pocedex.databinding.PokemonItemBinding

internal class PokemonListAdapter(private val pokemons: List<Pokemon>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var clickListener: ItemClickListener? = null

    private val all = 0
    private val favorite = 1

    override fun getItemViewType(position: Int): Int {
        return if (getPokemon(position).getSprite(0)==null) all else favorite
    }

    override fun onCreateViewHolder(parent: ViewGroup, ViewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return if (ViewType == all) {
            val binding = PokemonItemBinding.inflate(inflater, parent, false)
            ViewHolderAll(binding.root)
        } else {
            val binding = FavoriteGallaryBinding.inflate(inflater, parent, false)
            ViewHolderFavorite(binding.root)
        }

    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val pokemon = pokemons[position]
        if (holder.itemViewType == all) {
            val viewHolderAll: ViewHolderAll = holder as ViewHolderAll
            viewHolderAll.binding!!.pokemon = pokemon
        } else {
            val viewHolderFavorite: ViewHolderFavorite = holder as ViewHolderFavorite
            viewHolderFavorite.binding!!.pokemon = pokemon
        }
    }

    override fun getItemCount(): Int {
        return pokemons.size
    }

    inner class ViewHolderAll internal constructor(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        var binding: PokemonItemBinding? = null

        init {
            binding = DataBindingUtil.bind(view)

            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            if (clickListener != null) clickListener!!.onItemClick(view, adapterPosition)
        }
    }

    inner class ViewHolderFavorite internal constructor(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        var binding: FavoriteGallaryBinding? = null

        init {
            binding = DataBindingUtil.bind(view)

            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            if (clickListener != null) clickListener!!.onItemClick(view, adapterPosition)
        }
    }

    fun getPokemon(id: Int): Pokemon {
        return pokemons[id]
    }

    fun setClickListener(itemClickListener: ItemClickListener) {
        this.clickListener = itemClickListener
    }

    interface ItemClickListener {
        fun onItemClick(view: View, position: Int)
    }
}

